pkg/yurtctl/cmd/revert/edgenode.go

```diff
@@ -0,0 +1,182 @@
+			}
+			if err := r.RunRevertEdgeNode(); err != nil {
+				klog.Fatalf("fail to revert the yurt node to a kubernetes node: %s", err)
+			}
+		},
+	}
+	return cmd
+}
+
+func (r *RevertEdgeNodeOptions) Complete(flags *pflag.FlagSet) (err error) {
+	r.clientSet, err = kubeutil.GenClientSet(flags)
+	if err != nil {
+		return err
+	}
+
+	kubeletSvcPath := os.Getenv("KUBELET_SVC")
+	if kubeletSvcPath == "" {
+		kubeletSvcPath = enutil.KubeletSvcPath
+	}
+	r.kubeletSvcPath = kubeletSvcPath
+
+	openyurtDir := os.Getenv("OPENYURT_DIR")
+	if openyurtDir == "" {
+		openyurtDir = enutil.OpenyurtDir
+	}
+	r.openyurtDir = openyurtDir
+
+	podMainfestPath := os.Getenv("STATIC_POD_PATH")
+	if podMainfestPath == "" {
+		podMainfestPath = enutil.StaticPodPath
+	}
+	r.podMainfestPath = podMainfestPath
+
+	return
+}
+
+func (r *RevertEdgeNodeOptions) RunRevertEdgeNode() (err error) {
+	// 1. check the server version
+	if err = kubeutil.ValidateServerVersion(r.clientSet); err != nil {
+		return
+	}
+	klog.V(4).Info("the server version is valid")
+
+	// 2. check the state of worker node and remove label from node
+	nodeName, err := enutil.GetNodeName()
+	if err != nil {
+		return err
+	}
+	node, err := r.clientSet.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	isEdgeNode, ok := node.Labels[projectinfo.GetEdgeWorkerLabelKey()]
+	if ok && isEdgeNode == "true" {
+		// 2.1. check the state of worker nodes
+		_, condition := nodeutil.GetNodeCondition(&node.Status, v1.NodeReady)
+		if condition == nil || condition.Status != v1.ConditionTrue {
+			klog.Errorf("Cannot do the revert, the status of worker node: %s is not 'Ready'.", node.Name)
+			return
+		}
+
+		// 2.2. remove label from node
+		_, foundAutonomy := node.Annotations[constants.AnnotationAutonomy]
+		if foundAutonomy {
+			delete(node.Annotations, constants.AnnotationAutonomy)
+		}
+	}
+	if ok {
+		// remove the label for both the cloud node and the edge node
+		delete(node.Labels, projectinfo.GetEdgeWorkerLabelKey())
+		if _, err = r.clientSet.CoreV1().Nodes().Update(node); err != nil {
+			return
+		}
+	}
+	klog.Info("label alibabacloud.com/is-edge-worker is removed")
+
+	// 3. remove yurt-hub and revert kubelet service
+	if err := r.RevertKubelet(); err != nil {
+		return fmt.Errorf("fail to revert kubelet: %v", err)
+	}
+	if err := r.RemoveYurthub(); err != nil {
+		return err
+	}
+	return
+}
+
+func (r *RevertEdgeNodeOptions) RevertKubelet() error {
+	// 1. remove openyurt's kubelet.conf if exist
+	yurtKubeletConf := r.GetYurthubKubeletConf()
+	if err := os.Remove(yurtKubeletConf); err != nil {
+		return err
+	}
+	kubeletSvcBk := r.GetKubeletSvcBackup()
+	_, err := enutil.FileExists(kubeletSvcBk)
+	if err != nil {
+		klog.Errorf("fail to get file %s, will revise the %s directly", kubeletSvcBk, r.kubeletSvcPath)
+		return err
+	}
+	klog.Infof("found backup file %s, will use it to revert the node", kubeletSvcBk)
+	err = os.Rename(kubeletSvcBk, r.kubeletSvcPath)
```

Comment:
maybe we need delete the converted kubeletSvcPath file before rename the backup file?

Reply:
👌

Reply:
ok, i got it